icommands.c

```diff
@@ -98,6 +103,188 @@ enum CommandResult mutt_parse_icommand(/* const */ char *line, struct Buffer *er
+    return;
+  }
+
+  /* The pager and editor menus don't use the generic map,
+   * however for other menus try generic first. */
+  if ((menu != MENU_PAGER) && (menu != MENU_EDITOR) && (menu != MENU_GENERIC))
+  {
+    fn_name = mutt_get_func(OpGeneric, map->op);
+  }
+
+  /* if it's one of the menus above or generic doesn't find
+   * the function, try with it's own menu. */
+  if (!fn_name)
+  {
+    const struct Binding *bindings = km_get_table(menu);
+    if (!bindings)
+      return;
+
+    fn_name = mutt_get_func(bindings, map->op);
+  }
+
+  fprintf(fpout, "bind %s %s %s\n", menu_name, key_binding, fn_name);
+}
+
+/**
+ * bind_dump_menu - Dumps all the binds/macros of a menu into a file.
+ * @param fpout     Output file
+ * @param menu      Menu id to dump
+ * @param menu_name Name of the menu to dump
+ * @param bind      If true it's :bind ::icommand, else :macro
+ */
+static void dump_menu(FILE *fpout, const int menu, const char *menu_name, bool bind)
+{
+  struct Keymap *map = NULL, *next = NULL;
+
+  for (map = Keymaps[menu]; map; map = next)
+  {
+    next = map->next;
+
+    if (!bind && map->op == OP_MACRO)
+      dump_macro(fpout, menu_name, map);
+    else if (bind && map->op != OP_MACRO)
+      dump_bind(fpout, menu, menu_name, map);
+  }
+  fflush(fpout);
+}
+
+/**
+ * dump_all_menus - Dumps all the binds/macros inside of all menus
+ * @param fpout    Output file
+ * @param filename Name of the output file
+ * @param bind     If true it's :bind ::icommand, else :macro
+ */
+
+static void dump_all_menus(FILE *fpout, char *filename, bool bind)
+{
+  for (int i = 0; i < MENU_MAX; i++)
+  {
+    const char *menu_name = mutt_map_get_name(i, Menus);
+    const long file_size = mutt_file_get_size(filename);
+
+    dump_menu(fpout, i, menu_name, bind);
+
+    /* If initial size is different than actual size, menu wasn't
+    * empty. If so we add a newline for readability between menus. */
+    if (file_size != mutt_file_get_size(filename) && i < MENU_MAX - 1)
+    {
+      fputs("\n", fpout);
+      fflush(fpout);
+    }
+  }
+}
+
+/**
+ * icmd_bind - Parse 'bind' and 'macro' commands - Implements ::icommand_t
+ */
+static enum CommandResult icmd_bind(struct Buffer *buf, struct Buffer *s,
+                                    unsigned long data, struct Buffer *err)
+{
+  FILE *fpout = NULL;
+  char tempfile[PATH_MAX];
+  const bool bind = (data == 0);
+
+  if (!MoreArgs(s))
+    mutt_buffer_strcpy(buf, "all");
+  else
+    mutt_extract_token(buf, s, 0);
+
+  if (MoreArgs(s))
+  {
+    /* More arguments potentially means the user is using the
+     * ::command_t :bind command thus we delegate the task. */
+    return MUTT_CMD_ERROR;
+  }
+
+  mutt_mktemp(tempfile, sizeof(tempfile));
+  fpout = mutt_file_fopen(tempfile, "w");
+  if (!fpout)
+  {
+    mutt_buffer_printf(err, _("Could not create temporary file %s"), tempfile);
+    return MUTT_CMD_ERROR;
+  }
+
+  if (mutt_str_strcasecmp(buf->data, "all") == 0)
```

Comment:
... and test `bool all` to save a `strcmp()`

Reply:
But then we'll miss the case where the user types `bind all`, in which `all` may end being interpreted as a menu name instead of 'all menus'

Reply:
Sorry, I wasn't clear.
I meant to keep the `strcmp()`, but test the bool first.